### PR TITLE
TSV-794 Added form validation

### PR DESCRIPTION
**Pull request by jachro**

### CODE REVIEW

File: app/uk/gov/hmrc/formmappings/Mappings.scala [hunk 1]
```diff
@@ -25,9 +25,15 @@ object Mappings {
 
   def mandatoryText(error: => String): Mapping[String] =
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/formmappings/Mappings.scala]: This can return Mapping[NonEmptyString]

File: app/uk/gov/hmrc/formmappings/Mappings.scala [hunk 2]
```diff
@@ -25,9 +25,15 @@ object Mappings {
...
+      .transform[Option[String]](emptyStringAsNone, emptyStringAsNone)
       .verifying(error, _.isDefined)
       .transform[String](_.get, Some.apply)
 
+  private def emptyStringAsNone(maybeValue: Option[String]): Option[String] = maybeValue.map(_.trim).flatMap {
+    case "" => Option.empty[String]
+    case nonEmpty => Some(nonEmpty)
+  }
+
   def mandatoryLocalDate(formatError: => String): Mapping[LocalDate] =
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/formmappings/Mappings.scala]: Parameter name should be "errorKeyPrefix"

File: conf/messages [hunk 3]
```diff
@@ -5,15 +5,26 @@ personal-details.faded-heading=Confirm your identity
...
+personal-details.nino.required=Enter your National Insurance number.
+personal-details.nino.invalid=Enter your National Insurance number in the correct format. For example QQ123456C.
 personal-details.dateOfBirth=Date of birth
 personal-details.dateOfBirth.hint=For example, 31 3 1980
 personal-details.dateOfBirth.day=Day
 personal-details.dateOfBirth.month=Month
 personal-details.dateOfBirth.year=Year
+personal-details.dateOfBirth.invalid=Enter your date of birth in the correct format. For example, 31 3 1980.
+personal-details.dateOfBirth.day.required=Enter the day of the month from your date of birth.
+personal-details.dateOfBirth.day.invalid=The date you have entered is invalid. Enter a number between 1 and 31.
```
Comment (anshulbajpai) [inline conf/messages]: Check with Shital about this error message. Should it say day instead of date.

File: test/generators/Generators.scala [hunk 4]
```diff
@@ -41,15 +41,19 @@ trait Generators {
...
-  def strings(minLenght: Int, maxLength: Int): Gen[String] = for {
-    length <- Gen.chooseNum(minLenght, maxLength)
-    chars <- Gen.listOfN(length, Gen.alphaNumChar)
-  } yield chars.mkString
-
-  val nonEmptyStrings: Gen[String] = for {
-    length <- Gen.chooseNum(1, 1000)
-    chars <- Gen.listOfN(length, Gen.alphaNumChar)
-  } yield chars.mkString
+  def strings(minLenght: Int, maxLength: Int): Gen[String] = {
```
Comment (anshulbajpai) [inline test/generators/Generators.scala]: minLength spelling

File: test/generators/Generators.scala [hunk 5]
```diff
@@ -41,15 +41,19 @@ trait Generators {
...
+    for {
+      length <- Gen.chooseNum(minLenght, maxLength)
+      chars <- Gen.listOfN(length, Gen.alphaNumChar)
+    } yield chars.mkString
+  }.suchThat(_.trim.nonEmpty)
```
Comment (anshulbajpai) [inline test/generators/Generators.scala]: This will never produce a string of size 0 even if minLength parameter is 0

**Comment by anshulbajpai** [comment]
Show all form errors at the top of the form

File: test/uk/gov/hmrc/formmappings/MappingsSpec.scala [hunk 6]
```diff
@@ -41,101 +41,157 @@ class MappingsSpec
...
-      forAll(Gen.choose(1, 3), localDates) { (partsNumber, localDate) =>
+      forAll(Gen.oneOf(1, 2, 3), validDateParts) { (numberOfPartsToRemove, allParts) =>
 
-        val allParts = Seq(
-          "date.year" -> localDate.getYear.toString,
-          "date.month" -> localDate.getMonthValue.toString,
-          "date.day" -> localDate.getDayOfMonth.toString
-        )
-        val selectedParts = (1 to partsNumber).foldLeft(allParts) { (partsLeft, _) =>
+        val selectedParts = (1 to numberOfPartsToRemove).foldLeft(allParts) { (partsLeft, _) =>
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/formmappings/MappingsSpec.scala]: Extract the fold block into a method

File: test/uk/gov/hmrc/formmappings/MappingsSpec.scala [hunk 7]
```diff
@@ -201,10 +267,34 @@ class MappingsSpec
...
+
+    val toPartName: ((String, String)) => String = {
+      case (partName, _) => partName
+    }
+
+    def toErrorKeySuffixed(suffix: String): String => String =
+      name => s"error.key.$name.$suffix"
+
+    def toFormError(errorKey: String): FormError =
+      FormError("date", errorKey)
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/formmappings/MappingsSpec.scala]: Use variables for date field and errorKey

File: test/uk/gov/hmrc/formmappings/MappingsSpec.scala [hunk 8]
```diff
@@ -41,101 +41,157 @@ class MappingsSpec
...
-    "return the given error if date parts are invalid and there are additional constraints added" in {
+    "return the 'invalid' error if date parts forms invalid date" in new DateMappingSetup {
+
+      val partsWithInvalids = Map(
+        "date.year" -> "2017",
+        "date.month" -> "2",
+        "date.day" -> "29"
+      )
+
+      val bindResult = dateMapping.bind(partsWithInvalids.toMap)
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/formmappings/MappingsSpec.scala]: toMap is not required

File: app/uk/gov/hmrc/formmappings/LocalDateMapping.scala [hunk 9]
```diff
@@ -16,94 +16,162 @@
...
+
+    private implicit class ValidatedPartsOps(validatedParts: Seq[ValidatedNel[String, Int]]) {
 
-  private val toLocalDate: (Seq[Int]) => String Either LocalDate =
-    allParts =>
-      if (allParts.head < 1000)
-        Left(s"Year ${allParts.head} is invalid")
-      else Try {
-        LocalDate.of(allParts.head, allParts(1), allParts(2))
+      lazy val toValidatedDate: ValidatedNel[String, LocalDate] = validatedParts.foldLeft(Validated.validNel[String, Seq[Int]](Nil)) {
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/formmappings/LocalDateMapping.scala]: Try using sequence once you convert Validated to Either

File: app/uk/gov/hmrc/formmappings/LocalDateMapping.scala [hunk 10]
```diff
@@ -16,94 +16,162 @@
...
 
-  private val checkConstraints: LocalDate => Seq[FormError] Either LocalDate =
-    localDate =>
-      constraints.map(_.apply(localDate)).foldLeft(Seq.empty[FormError]) {
-        case (allErrors, Valid) => allErrors
-        case (allErrors, Invalid(errors)) => allErrors ++: errors.map(error => FormError(key, error.message, error.args))
-      } match {
-        case Nil => Right(localDate)
-        case validationErrors => Left(validationErrors)
+      private def toDate: Seq[Int] => ValidatedNel[String, LocalDate] = {
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/formmappings/LocalDateMapping.scala]: Use a tuple of 3 elements

File: app/uk/gov/hmrc/formmappings/LocalDateMapping.scala [hunk 11]
```diff
@@ -16,94 +16,162 @@
...
+}
+
+private object LocalDateMapping {
+
+  private case class LocalDateBinder(key: String,
+                                     constraints: Seq[Constraint[LocalDate]])
+                                    (errorKeyPrefix: => String) {
+
+    def bind(formData: Map[String, String]): Either[Seq[FormError], LocalDate] = Seq(
+      "year".prependWithKey.findValueIn(formData).parseToInt.validateUsing(YEAR.checkValidValue),
```
[the remark this answers is not in the document]
Comment (jachro) [reply app/uk/gov/hmrc/formmappings/LocalDateMapping.scala]: Once used with 'combine', Validated seems to be a better choice.

**Pull request merged**